### PR TITLE
Refactor CmdstanProcs

### LINKED ISSUE #197: Refactor CmdstanProcs
I started adding the generate_quantities method and this came up. Currently the CmdstanProcs is too sampling focused.

My current plan is to make a generic CmdstanProcs class with the common functionalities that are enough for VI/optimization/GQ and add a CmdstanMCMCProcs that inherits the generic one. But am open to other ideas.

This is currently blocking #7 We could fix it before this but it would just make this refactor more difficult.

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Fixes #197 

This PR does the following:
- everything in CmdstanProcs is now processes, not runs
- removed everything sampling specific out of CmdstanProcs 
- CmdstanMCMCProcs inherits everything from CmdstanProcs except for sampling two sampling specific functions
- all cmdstan runs use CmdstanProcs
- all fitting methods now print time on finish
- all fit objects now have fit$output()

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar, Univ. of Ljubljana

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
If the tests pass this is ready for review. I updated the top-level comment. This should now make generate_quantities much easier to implement.

**Comment by jgabry** [comment]
Great! I'll try to go through this today or this weekend.

File: R/run.R [hunk 1]
```diff
@@ -304,51 +310,50 @@ CmdStanRun$set("private", name = "run_variational_", value = .run_other)
...
+                                   .var.name = "parallel_procs", null.ok = TRUE)
+      checkmate::assert_integerish(threads_per_proc, lower = 1, len = 1, null.ok = TRUE,
+                                   .var.name = "threads_per_proc")
+      private$num_procs_ <- as.integer(num_procs)
+      if (is.null(parallel_procs)) {
+        private$parallel_procs_ <- private$num_procs_
+      } else {
+        private$parallel_procs_ <- as.integer(parallel_procs)
+      }
+      private$threads_per_proc_ <- threads_per_proc
```
Comment (jgabry) [inline R/run.R:336]: do we need `as.integer(threads_per_proc)`? or just for `parallel_procs`?

File: R/run.R [hunk 2]
```diff
@@ -396,26 +412,28 @@ CmdStanProcs <- R6::R6Class(
       private$total_time_ <- as.numeric(time)
       invisible(self)
     },
-    all_finished = function() {
-      finished <- TRUE
-      for (id in self$run_ids()) {
-        # if chain is not finished yet
+    check_finished = function() {
+      new_failed_procs = NULL
+      new_successful_procs = NULL
```
Comment (jgabry) [inline R/run.R:417]: I don't see `new_failed_procs` or `new_successful_procs ` used anywhere in `check_finished()`. They're returned at the end of the function but won't they always be `NULL`?

File: R/run.R [hunk 3]
```diff
@@ -396,26 +412,28 @@ CmdStanProcs <- R6::R6Class(
...
-            finished <- FALSE
+            self$mark_proc_stop(id)
+            self$report_time(id)            
           }
         }
       }
-      finished
+      list(
+        failed = new_failed_procs,
+        success = new_successful_procs
```
Comment (jgabry) [inline R/run.R:435]: See previous comment about these not being modified from `NULL`

File: R/run.R [hunk 4]
```diff
@@ -492,16 +507,53 @@ CmdStanProcs <- R6::R6Class(
...
       for (line in out) {
-        private$chain_output_[[id]] <- c(private$chain_output_[[id]], line)
+        private$proc_output_[[id]] <- c(private$proc_output_[[id]], line)
+        cat(line, collapse = "\n")
+      }
+      invisible(self)
+    },
+    report_time = function(id = NULL) {
+      cat("Finished in ",
+              format(round(mean(self$total_time()), 1), nsmall = 1),
```
Comment (jgabry) [inline R/run.R:522]: do we need `mean()` here? Isn't `self$total_time()` always a scalar?

File: R/run.R (hunk 1 above)
Comment (rok-cesnovar) [reply R/run.R:336]: Yep. Good call.

File: R/run.R (hunk 2 above)
Comment (rok-cesnovar) [reply R/run.R:417]: This is leftover from a previous implementation (I redid this a couple of times before settling on this). Will cleanup.

**Comment by rok-cesnovar** [comment]
Thanks @jgabry 
I addressed everything.

**Comment by jgabry** [comment]
Ok I just reverted using `as.integer()`. It was causing an error here: 

https://github.com/stan-dev/cmdstanr/blob/f9197e2c9e8c0af1ee9616e600e1d363e197862f/R/run.R#L224-L229

because it was never NULL so it always went into the `else` branch

**Comment by rok-cesnovar** [comment]
Thanks!

**Comment by jgabry** [comment]
All checks now passing! Merging now.

**Pull request merged**